pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,454 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud"
+	aliclient "github.com/gardener/gardener-extension-provider-alicloud/pkg/alicloud/client"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	ctrlerror "github.com/gardener/gardener/pkg/controllerutils/reconciler"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// bastionEndpoints holds the endpoints the bastion host provides
+type bastionEndpoints struct {
+	// private is the private endpoint of the bastion. It is required when opening a port on the worker node to allow SSH access from the bastion
+	private *corev1.LoadBalancerIngress
+	//  public is the public endpoint where the enduser connects to establish the SSH connection.
+	public *corev1.LoadBalancerIngress
+}
+
+// Ready returns true if both public and private interfaces each have either
+// an IP or a hostname or both.
+func (be *bastionEndpoints) Ready() bool {
+	return be != nil && IngressReady(be.private) && IngressReady(be.public)
+}
+
+func (a *actuator) Reconcile(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "reconcile")
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	credentials, err := alicloud.ReadCredentialsFromSecretRef(ctx, a.client, &opt.SecretReference)
+	if err != nil {
+		return err
+	}
+
+	aliCloudECSClient, err := a.newClientFactory.NewECSClient(opt.Region, credentials.AccessKeyID, credentials.AccessKeySecret)
+	if err != nil {
+		return err
+	}
+
+	aliCloudVPCClient, err := a.newClientFactory.NewVPCClient(opt.Region, credentials.AccessKeyID, credentials.AccessKeySecret)
+	if err != nil {
+		return err
+	}
+
+	imageID, err := getImageID(cluster, opt)
+	if err != nil {
+		return err
+	}
+
+	instanceType, err := aliCloudECSClient.GetInstanceType(cores)
+	if err != nil {
+		return err
+	}
+
+	if instanceType == nil ||
+		len(instanceType.AvailableZones.AvailableZone) == 0 ||
+		len(instanceType.AvailableZones.AvailableZone[0].AvailableResources.AvailableResource) == 0 ||
+		len(instanceType.AvailableZones.AvailableZone[0].AvailableResources.AvailableResource[0].SupportedResources.SupportedResource) == 0 ||
+		instanceType.AvailableZones.AvailableZone[0].AvailableResources.AvailableResource[0].SupportedResources.SupportedResource[0].Status != "Available" {
+		return errors.New("one cores instance is empty or unavailable")
```

Comment:
Currently it is hard-coded that we only use 1-core instance. However, should there really be circumstances where we do not have 1-core instance it would be pain in the back that we cannot create a bastion. Ideally we should be able to manually choose an instance type, this would provide additional option if the default one here doesn't work.